libvast/src/transform.cpp

```diff
@@ -109,57 +157,105 @@ transformation_engine::transformation_engine(std::vector<transform>&& transforms
 }
 
 /// Apply relevant transformations to the table slice.
-caf::expected<table_slice> transformation_engine::apply(table_slice&& x) const {
-  auto offset = x.offset();
-  auto size = x.rows();
+caf::error transformation_engine::add(table_slice&& x) {
+  VAST_DEBUG("transformation engine add");
+  range_.add(x.offset(), x.rows());
   auto layout = x.layout();
-  // TODO: Consider using a tsl robin map instead for transparent key lookup.
-  const auto& matching = layout_mapping_.find(std::string{layout.name()});
-  if (matching == layout_mapping_.end())
-    return std::move(x);
-  const auto& indices = matching->second;
-  VAST_INFO("applying {} transforms for received table slice w/ layout {}",
-            indices.size(), x.layout());
-  auto arrow_fast_path
-    = std::all_of(indices.begin(), indices.end(), [&](size_t idx) {
-        return transforms_.at(idx).arrow_fast_path_;
-      });
-  if (arrow_fast_path) {
-    VAST_INFO("selected fast path because all transforms support arrow");
-    // NOTE: It's important that `batch` is kept alive until `create()`
-    // is finished: If a copy was made, `batch` will hold the only reference
-    // to its underlying table slice, but the RecordBatches created by the
-    // transform step will most likely reference some of same underlying data.
-    auto original_batch = as_record_batch(x);
-    auto batch = original_batch;
+  to_transform_[layout].emplace_back(std::move(x));
+  return caf::none;
+}
+
+caf::error
+transformation_engine::process_queue(transform& transform, batch_queue& queue) {
+  caf::error failed{};
+  const auto size = queue.size();
+  for (size_t i = 0; i < size; ++i) {
+    auto [offset, layout, batch] = std::move(queue.front());
+    queue.pop_front();
+    auto add_failed = transform.add(offset, layout, batch);
+    if (add_failed) {
+      failed = add_failed;
+      while (!queue.empty())
+        queue.pop_front();
+      break;
+    }
+  }
+  // Finish frees up resource inside the plugin.
+  auto finished = transform.finish();
+  if (!finished && !failed)
+    failed = std::move(finished.error());
+  if (failed)
+    return failed;
+  for (const auto& b : *finished)
+    queue.push_back(b);
+  return caf::none;
+}
+
+/// Apply relevant transformations to the table slice.
+caf::expected<std::vector<table_slice>> transformation_engine::finish() {
+  VAST_DEBUG("transformation engine finish");
+  // NOTE: It's important that `batch` is kept alive until `create()`
+  // is finished: If a copy was made, `batch` will hold the only reference
+  // to its underlying table slice, but the RecordBatches created by the
+  // transform step will most likely reference some of same underlying data.
+  std::vector<std::shared_ptr<arrow::RecordBatch>> keep_alive{};
+  std::unordered_map<vast::type, batch_queue> batches{};
+  std::vector<table_slice> result{};
+  auto guard = caf::detail::make_scope_guard([this]() {
+    this->clear_slices();
+  });
```

Comment:
Why do we need manual memory management here? Can we use RAII somehow implicitly?

Reply:
This is the same problem as before:

One possible sequence of action: `new transform();` `add(slice1);` `add(slice2);` `finish();` `add(slice3);` `finish();` `delete transform`

The `transform.add()` has to store information related to the added slices. When `finish()` is called, these resources must be released before starting the next add-finish run.